File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Italy.m
//
//  UIImage+Italy.m
//  CountriesFlags
//

#import "Countries/UIImage+Italy.h"

@implementation UIImage (Italy)

+ (UIImage *)flagItalyWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:0.0f
                                          green:140.0f / 255.0f
                                           blue:69.0f / 255.0f
                                          alpha:1.0f];
    UIColor* colorWhite = [UIColor colorWithRed:244.0f / 255.0f
                                          green:245.0f / 255.0f
                                           blue:240.0f / 255.0f
                                          alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:205.0f / 255.0f
                                        green:33.0f / 255.0f
                                         blue:41.0f / 255.0f
                                        alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(size.width / 3.0f, 0.0f, size.width / 3.0f, size.height));

    [colorRed setFill];
    UIRectFill(CGRectMake(size.width / 3.0f * 2.0f, 0.0f, size.width / 3.0f, size.height));

    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Italy.swift
//
//  UIImage+Italy.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Italy "wikipedia"
     
     Generates the flag of Italy.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Italy.
     */
    internal static func flagItaly(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 0.0,
                                 green: 140.0/255.0,
                                 blue: 69.0/255.0,
                                 alpha: 1.0)
        let colorWhite = UIColor(red: 244.0/255.0,
                                 green: 245.0/255.0,
                                 blue: 240.0/255.0,
                                 alpha: 1.0)
        let colorRed = UIColor(red: 205.0/255.0,
                               green: 33.0/255.0,
                               blue: 41.0/255.0,
                               alpha: 1.0)
        
        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)
        
        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 3.0, y: 0.0),
                          size: CGSize(width: size.width / 3.0, height: size.height)))
        
        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: size.width / 3.0 * 2.0, y: 0.0),
                          size: CGSize(width: size.width / 3.0, height: size.height)))
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        
        return image
    }
}

#endif
